[ONFRONT-79] local add package

matt-stjean opened the pull request:
### Summary

This PR introduces support for the `--local` option in the `ontario-add-package` command. With this addition, users can now add local versions of Ontario packages to their projects using `npm link` instead of `npm install`.

### Changes

- **Local Package Linking**:
  - Added functionality to link local packages using `npm link` when the `--local` option is provided.
  - Updated the `handleAddPackageCommand` function to check for the `--local` option and link local packages accordingly.

- **Logging Enhancements**:
  - Improved logging throughout the `handleAddPackageCommand` function to provide better visibility into the process and facilitate debugging.

### Implementation Details

- **Link Local Packages**:
  - Added a `linkLocalPackage` function to handle the linking of local packages using `npm link`.
  - Adjusted the `handleAddPackageCommand` function to differentiate between third-party packages (installed via `npm install`) and local packages (linked via `npm link`).

- **Configuration Adjustments**:
  - Updated the package configuration to include separate arrays for `localPackages` and `thirdPartyPackages`, enabling precise handling of local package linking.

### Testing

- **Manual Testing**:
  - Verified that third-party packages are correctly installed using `npm install`.
  - Verified that local packages are correctly linked using `npm link` when the `--local` option is provided.
  - Ensured that logging outputs provide clear and detailed information on the package installation/linking process.

### Usage

To use the `--local` option with the `ontario-add-package` command:

```sh
npx ontario-add-package <package> --local
```

Example:

```sh
npx ontario-add-package eslint --local
```

### Related Issues

- Resolves ONFRONT-79: Support for `--local` option in `ontario-add-package`

matt-stjean committed:
feat(ontario-add-package): add support for local option

- Update jsdocs for affected methods
- Pass options to handler
- Handle options in handler

matt-stjean committed:
feat(ontario-add-package): adjust packages for local or remote installation

- Implement similar to handleCreateApp

matt-stjean committed:
chore(ontario-add-package): add more debugging

matt-stjean committed:
refactor(ontario-remove-package): handle package config change

- spread third party and local packages into one array
- pass new array to uninstall
- Update configFilesExist to remove instructions relevant only to addpackage

matt-stjean committed:
feat(ontario-add-package): support local option

- do not install local files
- link local files instead

matt-stjean committed:
[chore]: Update ontario-add-package.js fixing whitespace

matt-stjean committed:
chore(handleAddPackage): fix import of linklocal

syed-ods reviewed (commented):
Correction:
- The command `ontario-add-package eslint --local`, adds only ESLint and not the associated eslint-config package.
- The command `ontario-add-package prettier --local`, adds only Prettier and not the associated prettier-config package. 
- The command `ontario-remove-package <package-name> --local` works fine for both ESLint and Prettier. It also works fine without the `--local` flag. 

I also tried removing the core `ontario-frontend` npm package and install the local copy using `ontario-add-package ontario-frontend --local`, however I got the following error:
`error: command-argument value 'ontario-frontend' is invalid for argument 'package'. Allowed choices are eslint, prettier.`

I'm assuming the command is designed for the local ESLint and Prettier packages only.

kyle-binger-3 commented on packages/ontario-frontend-cli/core/utils/project/packageUtils.js:
```diff
@@ -64,14 +64,14 @@ async function isOntarioFrontendProject(dir = process.cwd()) {
  * @param {Array} configFiles - List of configuration files to check.
  * @returns {Promise<boolean>} - Returns true if any configuration file exists, otherwise false.
  */
-async function checkExistingConfigFiles(configFiles) {
+async function checkExistingConfigFiles(configFiles, noWarning) {
   let anyConfigFileExists = false;
   for (const { destination, warningMessage } of configFiles) {
     try {
       await fs.access(destination);
       anyConfigFileExists = true;
       logger.warning(
-        warningMessage || `Configuration file ${destination} already exists.`,
+        noWarning ? `${destination} exists.` : warningMessage
       );
     } catch (err) {
       // File does not exist, no action needed
```
Add the `noWarning` param to the JSDoc comment, it's unclear to me why it's needed, e.g. why would we want to suppress the warning?

kyle-binger-3 commented on packages/eslint-config-ontario-frontend/package.json:
```diff
@@ -20,11 +20,9 @@
   "author": "Ontario.ca Frontend",
   "license": "Apache-2.0",
   "main": "lib/.eslintrc.js",
-  "peerDependencies": {
-    "eslint": "^7.0.0 || ^8.0.0",
-    "eslint-plugin-import": "^2.0.1"
-  },
```
Out of scope but necessary, maybe just a quick note in the PR request / commit about this.

kyle-binger-3 commented on packages/ontario-frontend-cli/core/operations/npm/install.js:
```diff
@@ -33,8 +46,30 @@ async function installAllPackages(projectPath) {
+  ),
+  linkLocalPackages: withErrorHandling(
+    linkLocalPackages,
+    PackageInstallationError,
```
Does this need it's own error? e.g. `PackageLinkError`

matt-stjean committed:
fix(ontario-frontend-cli): handle conflicts with existing package installations and links

- Check for globally linked packages and unlink them to avoid conflicts.
- Check if the package is already linked locally and skip linking if found.
- Ensure local packages are linked correctly when the --local option is used.

Resolves potential conflicts and ensures a smooth linking process for local packages.

matt-stjean commented:
I have been spinning on the solution. Getting this to consistently work, covering edge cases, has been more effort than anticipated. This started as a way to save a bit of effort manually linking. 

Putting this in draft for the time being. 
@syed-ods @kyle-binger-3
